Hosted Agents Cache Volume Docs (#2737)
buildkite/docs: pages/pipelines/hosted_agents/linux.md

    </thead>
    <tbody>
        <tr><td>Small</td><td>2</td><td>4 GB</td></tr>

>         <tr><td>Medium</td><td>4</td><td>16 GB</td></tr>

Reviewer: From Oz, this is now 16 GB

<table data-attributes data-attributes-required>
  <tr>
    <td><code>paths</code></td>
    <td>
      A list of paths to cache. Paths are relative to the working directory of the step.<br>
      Absolute references can be provided in the cache paths configuration relative to the root of the instance.<br>
      <em>Example:</em><br>
      <code>- ".cache"</code><br>
      <code>- "/tmp/cache"</code><br>
    </td>
  </tr>
</table>

Optional attributes:

<table data-attributes data-attributes-required>
  <tr>
    <td><code>name</code></td>
    <td>
      A name for the cache. This allows for multiple cache volumes to be used in a single pipeline.<br>
      <em>Example:</em> <code>"node-modules-cache"</code><br>
    </td>
  </tr>

  <tr>
    <td><code>size</code></td>
    <td>
      The size of the cache volume. The default size is 20 gigabytes. Units are in gigabytes, specified as <code>Ng</code>, where <code>N</code> is the size in gigabytes, and <code>g</code> indicates gigabytes.<br>
      <em>Example:</em> <code>"20g"</code><br>
    </td>
  </tr>
</table>

### Lifecycle

At any point in time, multiple versions of a cache volume may be used by different jobs.

> The first request creates the first version of the cache volume, which is used as the parent of subsequent forks until a new parent version is committed.

Reviewer: What exactly is meant by "fork" in this context? I was thinking of a Git fork, but I can't see how cache volumes are stored in Git repos.

Reviewer: Fork of a storage volume 'moment'.

When requesting a cache volume, a "fork" of the previous cache volume version is attached to the agent instance (all but the first one, which starts empty).

> Each job gets its own private copy of the cache volume, as it existed at the time of the last cache commit.

Reviewer: Sorry, I'm having a little trouble following this. We might need a quick chat to clarify this.

Reviewer: Oz's response above has made this clearer.

## Cache volumes

> Cache Volumes are external volumes attached to hosted agent instances. Cache volumes are attached on best-effort basis depending on their locality, expiration and current usage (so they should not be relied upon as durable data storage).

Reviewer 2: ```suggestion
Cache Volumes are external volumes attached to hosted agent instances. Cache volumes are attached on a best-effort basis depending on their locality, expiration and current usage (so they should not be relied upon as durable data storage).
```

Reviewer 2: I'm assuming that we want to use Title Case for Cache Volumes as its a feature? https://github.com/buildkite/docs/blob/main/styleguides/writing-style.md#product-names-and-product-features

> Cache volumes are disabled by default and can be enabled by providing a list of paths to cache at the pipeline level or the step level. Cache volumes are scoped to a pipeline and are shared between all steps in a pipeline by default.

Reviewer 2: ```suggestion
Cache Volumes are disabled by default and can be enabled by providing a list of paths to cache at the pipeline level or the step level. Cache Volumes are scoped to a pipeline and are shared between all steps in a pipeline by default.
```

> Cache volumes act as regular disks with the following properties:

Reviewer 2: ```suggestion
Cache Volumes act as regular disks with the following properties:
```

- They're backed by local NVMe storage. You can expect high performance.

> - A Cache Volume is formatted as a regular Linux filesystem (e.g., ext4), so you can expect them to support any use-case you have that Linux supports.

Reviewer 2: ```suggestion
- A Cache Volume is formatted as a regular Linux filesystem (e.g. ext4), so you can expect them to support any use-case you have that Linux supports.
```

### Cache paths

> Cache volumes are mounted into user-specified paths. Cache paths will be mounted relative to the builds working directory. Absolute references can be provided in the cache paths configuration relative to the root of the instance (for example `/etc/cache`).

Reviewer 2: ```suggestion
Cache volumes are mounted into user-specified paths. Cache paths will be mounted relative to the build's working directory. Absolute references can be provided in the cache paths configuration relative to the root of the instance (for example, `/etc/cache`).
```

```yaml
# Mount the node_modules directory on all steps in the pipeline. This will use the default cache volume for the pipeline.
cache:

>   paths:

Reviewer 2: Do we need to specify somewhere that paths must be an Array?

    - "node_modules"

steps:
  - command: "npm install"

  - command: "bundle install"

>     # Mount the vendor/bundle directory to the cache volume.

Reviewer 2: ```suggestion
    # Mount the vendor/bundle directory to the Cache Volume.
```

>     # This will use the default cache volume for the pipeline.

Reviewer 2: ```suggestion
    # This will use the default Cache Volume for the pipeline.
```

    cache:
      paths:
        - "vendor/bundle"
```

### Lifecycle

At any point in time, multiple versions of a Cache Volume may be used by different jobs.

The first request creates the first version of the Cache Volume, which is used as the parent of subsequent forks until a new parent version is committed.

When requesting a Cache Volume, a "fork" of the previous cache volume version is attached to the agent instance (all but the first one, which starts empty).

Each job gets its own private copy of the Cache Volume, as it existed at the time of the last cache commit.

> Version commits follow a "last write" model: whenever a job terminates successfully (e.g. exits with exit code 0), Cache Volumes attached to that job have a new parent committed: the final flushed volume of the exiting agent instance.

Reviewer 2: ```suggestion
Version commits follow a "last write" model: whenever a job terminates successfully (i.e. exits with exit code 0), Cache Volumes attached to that job have a new parent committed: the final flushed volume of the exiting agent instance.
```

Whenever a job fails, the Cache Volume versions attached to the agent instance are abandoned.

### Custom caches

> Custom caches can be created by specifying a name for the cache. This allows for multiple cache volumes to be used in a single pipeline.

Reviewer 2: ```suggestion
Custom caches can be created by specifying a name for the cache. This allows for multiple Cache Volumes to be used in a single pipeline.
```

```yaml

> # Mount the node_modules directory on all steps. This will use the default cache volume for the pipeline.

Reviewer 2: ```suggestion
# Mount the node_modules directory on all steps. This will use the default Cache Volume for the pipeline.
```

cache:
  paths:
    - "node_modules"

steps:
  - command: "npm install"

  - command: "bundle install"

>     # Mount the vendor/bundle directory to a dedicated bundle cache volume. One bundle cache volume will be created for the pipeline.

Reviewer 2: ```suggestion
    # Mount the vendor/bundle directory to a dedicated bundle Cache Volume. One bundle Cache Volume will be created for the pipeline.
```

    cache:
      name: "bundle-cache"
      paths:
        - "vendor/bundle"
```

### Sizing

> When requesting a Cache Volume you can specify a size. When requesting x GB, a volume will be provided that has at least x GB free. In the case of a cache hit (most of the time), the actual volume size is: last used volume size + x.

Reviewer 2: ```suggestion
When requesting a Cache Volume, you can specify a size. The provided Cache Volume will have a minimum available storage equal to the specified size. In the case of a cache hit (most of the time), the actual volume size is: last used volume size + the specified size.
```

> The default size for the dependency cache is 20 gigabytes. This can be customized with the `size` option. Units are in gigabytes specified as a `Ng`, where N is the size in gigabytes.

Reviewer 2: ```suggestion
The default size for a Cache Volume is 20 gigabytes. This can be customized with the `size` option. Units are in gigabytes, specified as `Ng`, where N is the size in gigabytes.
```

Reviewer 2: I think its clearer to always refer to them as Cache Volumes rather than 'dependency cache' here

```yaml
cache:
  size: "30g"
  paths:
    - "node_modules"
steps:
  - command: "npm install"
```

### Git mirror cache

The Git mirror cache is a special type of Cache Volume that is used to speed up Git operations by caching the Git repository between builds. This is useful for large repositories that are slow to clone.

> Git mirror caching can be enabled on the Cluster's Cache Volumes settings page. Once enabled, the Git mirror cache will be used for all hosted jobs in that cluster. A separate cache volume will be created for each repository.

Reviewer 2: ```suggestion
Git mirror caching can be enabled on the cluster's Cache Volumes settings page. Once enabled, the Git mirror cache will be used for all hosted jobs in that cluster. A separate cache volume will be created for each repository.
```

## Cache volumes

_Cache volumes_ are external volumes attached to hosted agent instances. These volumes are attached on a best-effort basis depending on their locality, expiration and current usage, and therefore, should not be relied upon as durable data storage.

By default, cache volumes:

- Are disabled, although you can enable them by providing a list of paths to cache at the pipeline- or step-level.
- Are scoped to a pipeline and are shared between all steps in the pipeline.

Cache volumes act as regular disks with the following properties:

> - The volumes backed by local NVMe storage, delivering high performance.

Reviewer: What exactly is meant by "backed by" and where exactly is the "local" NVMe storage kept?

Does "backed by" mean use?

Reviewer: Means 'uses' in this context.

- The volumes are formatted as a regular Linux filesystem (e.g. ext4)—therefore, these volumes support any Linux use-cases.

### Cache configuration

Cache paths can be [defined in your `pipeline.yml`](/docs/pipelines/defining-steps) file. Defining cache paths for a step will implicitly create a cache volume for the pipeline.

> When cache paths are defined, the cache volume is mounted under `/cache` in the agent instance. The agent links subdirectories of the cache volume into the paths specified in the configuration. For example, defining `cache: "node_modules"` in your `pipeline.yml` file will link `./node_modules` to `/cache/bkcache/node_modules` in your agent instance.

Reviewer: Where does the intervening `/bkcache` directory come from?

I would have thought this would be `/cache/node_modules`?

Reviewer: Currently by design.

Reviewer: @matthewborden , quick question, would it be better to rewrite the first sentence to:

> When cache paths are defined, the cache volume is mounted under `/cache/bkcache` in the agent instance.

?

Author: `/cache/bkcache` should be relatively transparent to the customer, they should see that the volume is mounted into the directories they defined.

It would be more accurate to say:

> When cache paths are defined, the cache volume is mounted under /cache in the agent instance.

The `/bkcache` is created by the job setup scripts which will create matching directories for all the paths defined in `cache_paths` under `/cache/bkcache/user_defined_cache_path` then mount that directory into ./user_defined_cache_path within the checked out repo.

Custom caches can be created by specifying a name for the cache, which allows you to use multiple cache volumes in a single pipeline.

> When requesting a cache volume, you can specify a size. The cache volume provided will have a minimum available storage equal to the specified size. In the case of a cache hit (most of the time), the actual volume size is: last used volume size + the specified size.

Reviewer: What exactly is meant by a "cache hit"?

Reviewer: Means something is there in the cache.

Author: Yeah, it means you have received an existing cache that has been restored, rather than a new empty cache.
